Get rid of BehaviourPool & PlanPool

abhi1287 opened the pull request:
1. Get rid of BehaviourPool.
2. Life cycle of BasicBehaviour is tied to corresponding RunningPlan. This means multiple instances of a Behaviour could be active at any time regardless of configurations.
3. Deprecate support of configurations in Behaviours.
4. Simplify RunnableObject (called RunnableObjectNew for now) and BasicBehaviour,
5. Removed BasicBehaviour dependency on Scheduler. Alica engine thread would call init / terminate and timers will call run() methods.
6. Disabled execution order sensitive tests in this pr. Would re-enable them with the PlanPool pr. Reason: plan oninit and terminates are still being called by scheduler so things can get out of order.

abhi1287 committed:
Get rid of BehaviourPool

abhi1287 committed:
More cleanup

abhi1287 committed:
Further cleanup

abhi1287 committed:
Simplify construction of behaviour, new argument shouldm't require app code regeneration

abhi1287 committed:
Introduce PlanContext

abhi1287 committed:
Update alica tests

abhi1287 committed:
Update supplementary tests

abhi1287 committed:
Update alica test utility

abhi1287 committed:
Enable git lfs for jar files

abhi1287 committed:
Update readme

abhi1287 committed:
Rebase on top of cleanup#1

abhi1287 committed:
Working turtlesim - stress tested

abhi1287 committed:
Deprecate older init api

abhi1287 committed:
Merge rr_devel

abhi1287 committed:
Fix tests

abhi1287 committed:
Fix tests

abhi1287 committed:
Final test fixes

abhi1287 committed:
Format

abhi1287 committed:
Merge branch 'rr_devel' into abhi_mega_cleanup

abhi1287 committed:
cleanup

Dale-Koenig commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -33,88 +33,65 @@ struct BehaviourContext
  * The base class for all behaviours. All Behaviours must inherit from this class.
  */
-class BasicBehaviour : private RunnableObject
+class BasicBehaviour : private RunnableObjectNew
```
I guess this name is temporary until Planpool can use the same object?

abhi1287 replied:
Yes, will change in the 3rd cleanup pr after PlanPool.

Dale-Koenig commented on alica_engine/include/engine/RunnableObjectNew.h:
```diff
@@ -0,0 +1,131 @@
+    // Map from ConfAbstractPlanWrapper id to associated attachment
+    // Only plan will have these
+    std::unordered_map<int64_t, const KeyMapping*> _keyMappings;
+    TraceRunnableObject _runnableObjectTracer;
```
```suggestion
    TraceRunnableObject _runnableObjectTrace;
```

abhi1287 replied:
Its a tracer instance rather than trace. Just like NavigationTracer in LBC

Dale-Koenig commented on alica_engine/include/engine/RunnableObjectNew.h:
```diff
@@ -0,0 +1,131 @@
+    void stop();
+    void setEngine(AlicaEngine* engine) { _engine = engine; };
+    // Only plan will have these
+    void addKeyMapping(int64_t wrapperId, const KeyMapping* keyMapping);
```
Will these not end up in the derived class instead?

abhi1287 replied:
They will probably move to the RunningPlan. Will refactor in the 3rd cleanup pr.

Dale-Koenig commented on alica_test_utility/src/Util.cpp:
```diff
@@ -8,16 +8,27 @@ namespace alica::test
-                                      : behaviourEntry.first->getConfiguration()->getId() == configurationID)) {
-            behaviour = behaviourEntry.second.get();
-            break;
+    return getBasicBehaviourHelper(ae->getPlanBase().getRootNode(), behaviourID);
```
I guess later we should refactor these to only take the part of the engine we need

abhi1287 replied:
Could be in a separate pr, configuration is also unused.

Dale-Koenig reviewed (approved):
Not really qualified to judge all the details but definitely simpler than before

veerajsk commented on alica_engine/include/engine/RuntimeBehaviourFactory.h:
```diff
@@ -0,0 +1,33 @@
+class IAlicaWorldModel;
+
+/**
+ * Construct a runtime BasicBehaviour instance based
```
fix/complete this comment

abhi1287 replied:
done

veerajsk commented on alica_engine/src/engine/RuntimeBehaviourFactory.cpp:
```diff
@@ -0,0 +1,33 @@
+RuntimeBehaviourFactory::RuntimeBehaviourFactory(IBehaviourCreator& bc, IAlicaWorldModel* wm, AlicaEngine* engine)
+        : _creator(bc)
+        , _engine(engine)
+        , _wm(wm)
```
would be nice if the order here is same as the declaration order

veerajsk replied:
same for BasicBehaviour

veerajsk commented on alica_engine/include/engine/RunnableObjectNew.h:
```diff
@@ -0,0 +1,131 @@
+
+    RunnableObjectNew(IAlicaWorldModel* wm, const std::string& name = "");
+    virtual ~RunnableObjectNew() = default;
+    // virtual ~RunnableObjectNew() { std::cout << _name << " destructor" << std::endl; }
```
remove

abhi1287 replied:
done

veerajsk commented on alica_engine/src/engine/RunnableObjectNew.cpp:
```diff
@@ -0,0 +1,189 @@
+
+void RunnableObjectNew::setupBlackboard()
+{
+    if (!_runningplanContext->getParent() || !_runningplanContext->getParent()->getBasicPlan()) {
```
what is the second condition for? Shouldn't it always return true (except for master plan which is already verified in the first condition)?

abhi1287 replied:
a RunningPlan could also represent PlanType which temporarily might not have a BasicPlan assigned yet.

abhi1287 replied:
Anyways this should move out to RunningPlan

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -44,134 +50,70 @@ AgentId BasicBehaviour::getOwnId() const
-        _activeRunJobId = -1;
-    }
     // Just to be double safe in terms of the correct behaviour of isSuccess() & isFailure() ensure result is reset before incrementing _execState
     _behResult.store(BehResult::UNKNOWN);
```
with this refactor, this line should be moved after calling onTermination because there is nothing preventing onTermination() from setting success/failure, which would mean that isSuccess() & isFailure() will return invalid & stale values for the next run

abhi1287 replied:
Object would be deleted after onTermination so there won't be any next run. But yeah logically it makes sense to reset after calling all domain functions.

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -44,134 +50,70 @@ AgentId BasicBehaviour::getOwnId() const
-        return;
-    }
 
     // Intentionally call onTermination() at the end. This prevents setting success/failure from this method
```
this is no longer applicable & I don't see any harm in setting success/failure in onTermination now

abhi1287 replied:
there is no point in setting success or failure in ontermination as engine has already decided to shutdown the behaviour. Better utility is for isSuccess() and isFailure() methods. This information would be available in onTermination method now.

veerajsk commented on alica_tests/src/test/Expr/src/Behaviour/ReadConfigurationBehaviour.cpp:
```diff
@@ -34,7 +34,8 @@ void ReadConfigurationBehaviour::run(void* msg)
     /*PROTECTED REGION ID(initialiseParameters1588061129360) ENABLED START*/
-    getParameter("TestValue", testValue);
+    // getPlanContext()->getParameter("TestValue", testValue);
+    testValue = "1";
```
can't we get rid of this behaviour since we no longer have configurations

abhi1287 replied:
Configuration cleanup will follow in a later cleanup pr.

veerajsk commented on alica_engine/src/engine/RunningPlan.cpp:
```diff
@@ -480,7 +480,7 @@ void RunningPlan::deactivate()
 
     if (isBehaviour()) {
-        _ae->editBehaviourPool().stopBehaviour(*this);
+        _basicBehaviour->stop();
```
I think we should also change how RunningPlan's are destroyed. Right now, a vector of RunningPlan shared_ptrs are stored in the PlanBase which is iterated over to find RunningPlan's that can be deleted by calling RunningPlan::isDeletable. One reason why this was required earlier is because the behaviours attached to a running plan could still be running even if we transitioned out of it, but with this refactor this won't be the case. isDeletable does check a few more things but I think we should be able to destroy the RunningPlan when it is deactivated.

abhi1287 replied:
Yes, would do it in the 3rd cleanup pr

abhi1287 committed:
Merge branch 'rr_devel' into abhi_mega_cleanup

abhi1287 committed:
Review comments fixes

abhi1287 committed:
Remove PlanPool (#259)

* Remove PlanPool

* Fix reactivation of plan

- Ignore duplicate calls to start & stop in RunnableObject
- Stop the basicPlan on reactivation & allocate a new BasicPlan object
if the plan has changed

Co-authored-by: Veeraj S Khokale <veeraj.khokale@rapyuta-robotics.com>

veerajsk committed:
Address review comment
